[JSC] Emit logical op children in condition context when overall result is ignored

ddegazio opened the pull request:
#### eeecca442eb0e974f22a45b07660754a42106275
<pre>
[JSC] Emit logical op children in condition context when overall result is ignored
<a href="https://bugs.webkit.org/show_bug.cgi?id=289134">https://bugs.webkit.org/show_bug.cgi?id=289134</a>
<a href="https://rdar.apple.com/146174665">rdar://146174665</a>

Reviewed by Yusuke Suzuki and Geoffrey Garen.

Emits logical op children in condition context, meaning we compile them down
to conditional branches instead of getting their values, when the overall
result of a logical operation is known to be ignored. This fixes an issue
where we were unable to fuse conditions and branches in discarded logical
operations because the result value was assumed to be accessible. Also changes
SourceElements to ignore the results of child statements unless the completion
value is observable.

* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::LogicalOpNode::emitBytecode):
(JSC::SourceElements::emitBytecode):

Canonical link: <a href="https://commits.webkit.org/291740@main">https://commits.webkit.org/291740@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0e61496c6505a1a47e9fba542213e9ba7af1d4e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93775 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13349 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3087 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98781 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44301 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13645 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21791 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71583 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28954 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96777 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10151 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84746 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51917 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9833 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2378 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43616 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86485 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2448 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100816 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92441 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20827 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15198 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80597 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21079 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80694 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79938 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19912 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24493 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1845 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13984 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20811 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25989 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115091 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20498 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23958 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22239 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
r=me

This seems fine, but when I checked whether we could do the same optimization for other nodes, I also found `emitNodeInIgnoreResultPosition`. What's our thinking for when the caller should call `emitNodeInIgnoreResultPosition` vs when the callee should check `dst == ignoredResult()`?

ddegazio committed:
[JSC] Emit logical op children in condition context when overall result is ignored
https://bugs.webkit.org/show_bug.cgi?id=289134
rdar://146174665

Reviewed by Yusuke Suzuki and Geoffrey Garen.

Emits logical op children in condition context, meaning we compile them down
to conditional branches instead of getting their values, when the overall
result of a logical operation is known to be ignored. This fixes an issue
where we were unable to fuse conditions and branches in discarded logical
operations because the result value was assumed to be accessible. Also changes
SourceElements to ignore the results of child statements unless the completion
value is observable.

* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::LogicalOpNode::emitBytecode):
(JSC::SourceElements::emitBytecode):

Canonical link: https://commits.webkit.org/291740@main
